google/cloud/aiplatform/models.py

```diff
@@ -215,10 +228,534 @@ def upload(
+                will be scaled down to accommodate new deployed model's traffic.
+                Should not be provided if traffic_split is provided.
+            traffic_split (Dict[str, int]):
+                Optional. A map from a DeployedModel's ID to the percentage of
+                this Endpoint's traffic that should be forwarded to that DeployedModel.
+                If a DeployedModel's ID is not listed in this map, then it receives
+                no traffic. Key for model being deployed is "0". Should not be
+                provided if traffic_percentage is provided.
+            machine_type (str):
+                Optional. The type of machine. Not specifying machine type will
+                result in model to be deployed with automatic resources.
+            min_replica_count (int):
+                Optional. The minimum number of machine replicas this deployed
+                model will be always deployed on. If traffic against it increases,
+                it may dynamically be deployed onto more replicas, and as traffic
+                decreases, some of these extra replicas may be freed.
+            max_replica_count (int):
+                Optional. The maximum number of replicas this deployed model may
+                be deployed on when the traffic against it increases. If requested
+                value is too large, the deployment will error, but if deployment
+                succeeds then the ability to scale the model to that many replicas
+                is guaranteed (barring service outages). If traffic against the
+                deployed model increases beyond what its replicas at maximum may
+                handle, a portion of the traffic will be dropped. If this value
+                is not provided, the smaller value of min_replica_count or 1 will
+                be used.
+            metadata (Sequence[Tuple[str, str]]):
+                Optional. Strings which should be sent along with the request as
+                metadata.
+        Returns:
+            endpoint ("Endpoint"):
+                Endpoint with the deployed model.
+
+        """
+        if endpoint is None:
+            display_name = self.display_name + "_endpoint"
+            endpoint = Endpoint.create(display_name=display_name)
+
+        endpoint.deploy(
+            model=self,
+            deployed_model_display_name=deployed_model_display_name,
+            traffic_percentage=traffic_percentage,
+            traffic_split=traffic_split,
+            machine_type=machine_type,
+            min_replica_count=min_replica_count,
+            max_replica_count=max_replica_count,
+            metadata=metadata,
+        )
+
+        return endpoint
+
+
+class Endpoint(base.AiPlatformResourceNoun):
+
+    client_class = EndpointServiceClient
+    _is_client_prediction_client = False
+
+    def __init__(
+        self,
+        endpoint_name: Optional[gca_endpoint.Endpoint] = None,
```

Comment:
I understand this was set as Optional to be able to create the class while the operations future was still executing. The issue I see is that the user will see this and think that they can get an endpoint with no resource name.

1. I think it's safer in this PR to require this argument and block during create and instantiate the class after the LRO is complete (so no need to pass in None).

2. In the future when we provide async support we should rely on an alternative constructor to initialize the endpoint to None so we don't surface that implementation detail on our APIs surface to users. (doesn't need to be added to this PR).

Reply:
will block for now

Reply:
Also, this should be Optional[str].